internal/app/cache/cache.go

```diff
@@ -11,9 +17,130 @@ type Cache interface {
 	// Fetch returns the cached response if it exists.
 	Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error)
 
-	// SetResponse sets the cache response and returns the list of open watches.
+	// SetResponse sets the cache response and returns the list of requests.
 	SetResponse(key string, resp envoy_api_v2.DiscoveryResponse) ([]*envoy_api_v2.DiscoveryRequest, error)
 
-	// AddWatch adds the watch to the cache and returns whether a stream is open.
-	AddWatch(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
+	// AddRequest adds the request to the cache and returns whether a stream is open.
+	AddRequest(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
+}
+
+type cache struct {
+	cacheMu       sync.RWMutex
+	cache         *ristretto.Cache
+	expireSeconds time.Duration
+}
+
+type resource struct {
+	resp       *envoy_api_v2.DiscoveryResponse
+	requests   []*envoy_api_v2.DiscoveryRequest
+	streamOpen bool
+}
+
+func NewCache(numCounters int64, cacheSizeBytes int64, expireSeconds int, onEvict func(key, conflict uint64,
+	value interface{}, cost int64)) (Cache, error) {
+	// Config values are set as recommended in the ristretto documentation: https://github.com/dgraph-io/ristretto#Config.
+	config := ristretto.Config{
+		// NumCounters sets the number of counters/keys to keep for tracking access frequency.
+		// The value is recommended to be higher than the max cache capacity,
+		// and 10x the number of unique items in the cache when full (note that this is different from MaxCost).
+		NumCounters: numCounters,
+		// MaxCost denotes the max size in bytes and determines how eviction decisions are made.
+		// For consistency, the cost passed into SetWithTTL represents the size of the value in bytes.
+		MaxCost: cacheSizeBytes,
+		// BufferItems is the size of the Get buffers. The recommended value is 64.
+		BufferItems: 64,
+		// OnEvict is called for each eviction and closes the stream if a key is removed (e.g. expiry due to TTL).
+		OnEvict: onEvict,
+	}
+	newCache, err := ristretto.NewCache(&config)
+	if err != nil {
+		return nil, err
+	}
+	return &cache{
+		cache:         newCache,
+		expireSeconds: time.Duration(expireSeconds) * time.Second,
+	}, nil
+}
+
+func (c *cache) Exists(key string) bool {
+	c.cacheMu.RLock()
+	_, found := c.cache.Get(key)
+	c.cacheMu.RUnlock()
+	return found
+}
+
+func (c *cache) Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error) {
+	c.cacheMu.RLock()
```

Comment:
What's the reasoning behind locking on reads?

Reply:
I was concerned about reads to the cache kicking off multiple writes if AddRequest/SetResponse hadn't finished, but I don't think the locking on reads is needed for Exists/Fetch.